packages/workspace/src/workspace/static_files/source.ts

```diff
@@ -42,34 +42,31 @@ class MissingStaticFileError extends Error {
 }
 
 export class AbsoluteStaticFile extends StaticFile {
-  protected dirStore: DirectoryStore<Buffer>
-
   constructor(
     params: StaticFileParameters,
-    dirStore: DirectoryStore<Buffer>,
+    readonly absoluteFilePath: string,
   ) {
     super(params)
-    this.dirStore = dirStore
-  }
-
-  get absoluteFilePath(): string {
-    return this.dirStore.getFullPath(this.filepath)
   }
 }
 
 export class LazyStaticFile extends AbsoluteStaticFile {
+  private didRead = false
+
   constructor(
     filepath: string,
     hash: string,
-    dirStore: DirectoryStore<Buffer>,
+    absoluteFilePath: string,
+    private getContentFunc: () => Promise<Buffer | undefined>,
     encoding?: BufferEncoding
   ) {
-    super({ filepath, hash, encoding }, dirStore)
+    super({ filepath, hash, encoding }, absoluteFilePath)
   }
 
   async getContent(): Promise<Buffer | undefined> {
-    if (this.internalContent === undefined) {
-      this.internalContent = (await this.dirStore.get(this.filepath))?.buffer
+    if (!this.didRead) {
+      this.internalContent = (await this.getContentFunc())
+      this.didRead = true
```

Comment:
```suggestion
    if (this.resolvedContent === undefined) {
      this.resolvedContent = this.getContentFunc()
    }
    return this.resolvedContent
```

with the current implementation we can read the content an unlimited amount of times (`didRead` is only updated to `true` after the `await`, so all calls until the content is resolved will see `didRead = false` and read again).

I think `internalContent` isn't actually needed here, right?
I think it isn't needed at all outside the base class, maybe it should be changed to be private?